src/storage-client/src/storage_collections.rs

```diff
@@ -0,0 +1,2816 @@
+                    diagnostics.clone(),
+                )
+                .await
+                .expect("invalid persist usage");
+
+            // Take the join of the handle's since and the provided `since`;
+            // this lets materialized views express the since at which their
+            // read handles "start."
+            let since = handle
+                .since()
+                .join(since.unwrap_or(&Antichain::from_elem(T::minimum())));
+
+            let our_epoch = self.envd_epoch;
+
+            loop {
+                let current_epoch: PersistEpoch = handle.opaque().clone();
+
+                // Ensure the current epoch is <= our epoch.
+                let unchecked_success = current_epoch.0.map(|e| e <= our_epoch).unwrap_or(true);
+
+                if unchecked_success {
+                    // Update the handle's state so that it is in terms of our
+                    // epoch.
+                    let checked_success = handle
+                        .compare_and_downgrade_since(
+                            &current_epoch,
+                            (&PersistEpoch::from(our_epoch), &since),
+                        )
+                        .await
+                        .is_ok();
+                    if checked_success {
+                        break handle;
+                    }
+                } else {
+                    mz_ore::halt!("fenced by envd @ {current_epoch:?}. ours = {our_epoch}");
+                }
+            }
+        };
+
+        let mut write = persist_client
+            .open_writer(
+                shard,
+                Arc::new(relation_desc),
+                Arc::new(UnitSchema),
+                diagnostics.clone(),
+            )
+            .await
+            .expect("invalid persist usage");
+
+        // N.B.
+        // Fetch the most recent upper for the write handle. Otherwise, this may
+        // be behind the since of the since handle. Its vital this happens AFTER
+        // we create the since handle as it needs to be linearized with that
+        // operation. It may be true that creating the write handle after the
+        // since handle already ensures this, but we do this out of an abundance
+        // of caution.
+        //
+        // Note that this returns the upper, but also sets it on the handle to
+        // be fetched later.
+        write.fetch_recent_upper().await;
+
+        (write, since_handle)
+    }
+
+    fn register_handles(
+        &self,
+        id: GlobalId,
+        is_in_txns: bool,
+        since_handle: SinceHandle<SourceData, (), T, Diff, PersistEpoch>,
+        write_handle: WriteHandle<SourceData, (), T, Diff>,
+    ) {
+        self.send(BackgroundCmd::Register {
+            id,
+            is_in_txns,
+            since_handle,
+            write_handle,
+        });
+    }
+
+    fn send(&self, cmd: BackgroundCmd<T>) {
+        self.cmd_tx.send(cmd).expect("task unexpectedly shut down");
```

Comment:
I've never seen this panic, so isn't it better to be sure that the send went through. As far as I see, this should never fail because `&self` here holds a handle to the task, so it can't be shutting down. Wdyt?

Reply:
It still possible, based on the timing of the runtime shutdown; we don't really do clean shutdowns ever, but I recommend loosening this, we have seen code exactly like this cause noise in storage before!

Reply:
will do then!